source/material_model/rheology/compositional_viscosity_prefactors.cc

```diff
@@ -0,0 +1,169 @@
+/*
+  Copyright (C) 2024 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/material_model/rheology/compositional_viscosity_prefactors.h>
+#include <aspect/utilities.h>
+#include <aspect/global.h>
+#include <aspect/geometry_model/interface.h>
+
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/base/parameter_handler.h>
+#include <aspect/simulator_signals.h>
+#include <aspect/parameters.h>
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    namespace Rheology
+    {
+      template <int dim>
+      CompositionalViscosityPrefactors<dim>::CompositionalViscosityPrefactors ()
+        = default;
+
+      template <int dim>
+      std::vector<double>
+      CompositionalViscosityPrefactors<dim>::compute_viscosities (const MaterialModel::MaterialModelInputs<dim> &in,
+                                                                  const double base_viscosity,
+                                                                  const unsigned int composition_index,
+                                                                  const unsigned int q) const
+      {
+        std::vector<double> factored_viscosities(number_of_prefactors, base_viscosity);
+        switch (viscosity_prefactor_scheme)
+          {
+            case none:
+            {
+              break;
+            }
+            case hk04_olivine_hydration:
+            {
+              // We use the atomic H_Si_ratio_ppm composition (C_OH) at each point to compute the water fugacity
+              // of olivine assuming a composition of 90 mol% Forsterite and 10 mol% Fayalite from Hirth and
+              // Kohlstaedt 2004 10.1029/138GM06.
+              const unsigned int bound_fluid_idx = this->introspection().compositional_index_for_name("bound_fluid");
+              const double weight_fraction_H2O = in.composition[q][bound_fluid_idx]; // mass fraction of bound water
```

Comment:
Perhaps make the terminology in the comment (mass) consistent with the variable name? They are synonymous here (I think) due to the fact that this is a percentage, but it still might be good to be consistent to avoid any confusion later on.